### ISSUE TITLE
PERF401 (use-list-comprehension) with walrus operator

### ISSUE THREAD

**Issue by fabien-michel** [astral-sh/ruff#6056]
I'm not sure the PERF401 should trigger when a walrus operator is implied in the condition.

The warning appear for this code.
```python
def odd_only(a):
    if a % 2:
        return a
    return False

filtered = []
for a in range(1, 5):
    if b := odd_only(a):
        filtered.append(b)
```
Which should be changed to
```python
filtered = [b for a in range(1, 5) if (b := odd_only(a))]
```
I'm found this kind of code not clear, because of the walrus operator.

The workaround is to not use the walrus operator:
```python
filtered = []
for a in range(1, 5):
    b = odd_only(a)
    if b:
        filtered.append(b)
```
So, as there is an equivalence, and the warn is not triggered, it should also not warn when a walrus operator is used.

The same behavior occur with perflint, I submitted the same bug: https://github.com/tonybaloney/perflint/issues/43

**Comment by tjkuson** [comment]
I am not sure `use-list-comprehension` should ignore the walrus operator; my understanding of [PEP 572](https://peps.python.org/pep-0572/) is that use in comprehensions is an explicit motivation for the assignment operator.

**Comment by fabien-michel** [comment]
You are true, sorry for the noise. Personally, I find this type of writing difficult to read.

**Comment by tjkuson** [comment]
No need to apologise! Many Python developers don't like the walrus operator. Personally, I like them. However, there's probably a compelling reason to implement something like [flake8-walrus](https://github.com/asottile/flake8-walrus) in a hypothetical Ruff category that prevents the use of language and library features (similar to `clippy::restrictions`).
